Wangscape/utils.h

```diff
@@ -0,0 +1,64 @@
+#pragma once
+#include <type_traits>
+#include <stdexcept>
+
+namespace
+{
+
+template<typename I>
+I _pow(I base, I exp)
+{
+    I result = 1;
+    for (int i = 0; i < std::numeric_limits<I>::digits; i++)
+    {
+        if (exp & 1)
+            result *= base;
+        else if (exp == 0)
+            break;
+        exp >>= 1;
+        base *= base;
+    }
+    return result;
+}
+
+}
+
+template <typename I>
+typename std::enable_if<std::is_unsigned<I>::value, I>::type pow(I base, I exp)
+{
+    if (exp == 0)
+        return 1;
+    else if (base == 0)
+        return 0;
+    else if (base == 1)
+        return 1;
```

Comment:
Regarding extracting common checks into functions: the sequence of conditions where the checks must be made is:
```
(exp == 0) // all
else if (base == 1) // all
else if (base == -1) // base signed
else if (exp < 0) // exp signed
else if (base == 0) // all
else if (exp > std::numeric_limits<I>::digits) // base signed
else if (base == -2) // base signed
else if (exp >= std::numeric_limits<I>::digits) // all
else if (base == 2) // all
```
With the current template setup (`I base, I exp`), separating out all of the common checks would mean splitting the function into 5 pieces. With `IBase base, IExp exp`, it would require 6. I tried to do this pairs of `enable_if` functions and the result was rather long and not very readable. If I use the current method with `IBase base, IExp exp`, it will require 2 more specialisations with a lot of duplicated code.

I've made a version with [static_if](https://github.com/wichtounet/cpp_utils/blob/master/static_if.hpp) (included as a submodule), which keeps everything in one function and is much more readable.